### PR TITLE
graalvm-ce: multiple fixes and improvements

### LINKED ISSUE #142392: graalvm-ce: `native-image --static --libc=musl` doesn't work
### Describe the bug

Usage of `--static` flag without `--libc=musl` [is not documented by upstream](https://www.graalvm.org/reference-manual/native-image/StaticImages/), and it is known in the Clojure community that its usage causes issues. This is because without `--libc=musl` the binaries are statically linked with `glibc`, and `glibc` [strongly discourages statically linking.](https://sourceware.org/glibc/wiki/FAQ#Even_statically_linked_programs_need_some_shared_libraries_which_is_not_acceptable_for_me.__What_can_I_do.3F)

I tried to create a GraalVM derivation compatible with `musl`. The result is here: https://github.com/thiagokokada/graalvm-nix. However, this can't be upstreamed since it brokes all compilations with `glibc` (including dynamically linked programs as used in nixpkgs).

The major issue is that, to make compilation work, [we symbolic link all libraries to the GraalVM closure](https://github.com/NixOS/nixpkgs/blob/8abb6e72c91ef08847dca4f1f35d0e76d1b7c338/pkgs/development/compilers/graalvm/community-edition/mkGraal.nix#L120-L125). This approach works fine for when we have only one lib C, however if we have two separate lib C we will have name conflicts and either we will have a compilation with glibc symbols on musl libraries or vice-versa.

So, what are the options? One approach I can think is extend the wrapper from `native-image` binary to conditionally include either `glibc` or `musl` depending on the context (e.g.: if the `--libc=musl` flag is passed or not). We can do this using the `native-image -H:CLibraryPath` flag, as below:

https://github.com/NixOS/nixpkgs/blob/e7b3740842a3f82c9235d0dc48867653d2a73c96/pkgs/development/interpreters/clojure/babashka.nix#L26

For this to work we also need to remove the symbolic links to glibc library, otherwise this can conflict with musl libraries. This should also fix this [issue](https://github.com/NixOS/nixpkgs/issues/140544#issuecomment-934211482). However, this looks like a very ugly solution, and I am open for better ones.

### Steps To Reproduce

0. `nix-shell -p graalvm11-ce musl.dev`. Keep in mind after #142243, it expects `musl-gcc` binary as `x86_64-linux-musl-gcc`. You can create a symbolic link and add it to your `PATH`
1. Create a file `HelloWorld.java` with contents:
    ```java
    public class HelloWorld {
      public static void main(String[] args) {
        System.out.println("Hello World");
      }
    }
    ```
2. `javac HelloWorld.java`
3. `native-image --no-server --libc=musl --static HelloWorld`
4. `./helloworld` (segfaults)

### Expected behavior
It should compile and run successfully.

### Screenshots
N/A

### Notify maintainers

@ericdallo @hlolli @babariviere @glittershark @bennyandresen

### Metadata
Please run `-` and paste the result.

```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 5.10.71, NixOS, 21.05.20211015.83667ff (Okapi)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.4pre20211006_53e4794`
 - nixpkgs: `/nix/store/pq6lkx0i52lqbyvnhiwzar7mmc907vcf-source`
```

Maintainer information:
```yaml
# a list of nixpkgs attributes affected by the problem
attribute: graalvm11-ce
# a list of nixos modules affected by the problem
module:
```

### PR DESCRIPTION
**Pull request by thiagokokada**
###### Motivation for this change

Sync improvements from my https://github.com/thiagokokada/graalvm-nix repository:

- Remove some unnecessary macOS dependencies (e.g.: GTK3 support)
- Patch openssl.so's Ruby libraries so autoPatchelfIgnoreMissingDeps is not needed anymore on Linux
- Add support for musl (closes https://github.com/NixOS/nixpkgs/issues/142392, however in a non-optimal way)
- Add binutils/stdenv.cc to runtime, making this derivation pure   (it used to depend on GCC installed in system)
- Format file with nixpkgs-fmt

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by thiagokokada** [comment]
@ofborg build graalvm11-ce graalvm17-ce babashka

**Comment by thiagokokada** [comment]
@ofborg build graalvm11-ce graalvm17-ce babashka

File: pkgs/development/compilers/graalvm/community-edition/mkGraal.nix [hunk 1]
```diff
@@ -108,55 +139,73 @@ let
...
-    '';
 
     dontStrip = true;
 
-    preFixup = ''
+    # Workaround for libssl.so.10 wanted by TruffleRuby
+    # Resulting TruffleRuby cannot use `openssl` library.
+    autoPatchelfIgnoreMissingDeps = stdenv.isDarwin;
+
+    preFixup = lib.optionalString (stdenv.isLinux) ''
```
Comment (thiagokokada) [inline pkgs/development/compilers/graalvm/community-edition/mkGraal.nix:208]: `patchelf` only exists on Linux.

**Comment by thiagokokada** [comment]
@ofborg build graalvm11-ce babashka

For aarch64-darwin.

**Comment by thiagokokada** [comment]
Looks fine now, builds on all platforms (forgot that `aarch64-darwin` is not a valid platform yet).

CC @ericdallo

**Comment by thiagokokada** [comment]
Result of `nixpkgs-review pr 152379` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>11 packages built:</summary>
  <ul>
    <li>babashka</li>
    <li>clj-kondo</li>
    <li>clojure-lsp</li>
    <li>dapl-native</li>
    <li>dbqn-native</li>
    <li>graalvm11-ce</li>
    <li>graalvm17-ce</li>
    <li>jet</li>
    <li>neil</li>
    <li>vscode-extensions.betterthantomorrow.calva</li>
    <li>zprint</li>
  </ul>
</details>

**Review by ericdallo — APPROVED** [review]
Looks good!

**Comment by thiagokokada** [comment]
Result of `nixpkgs-review pr 152379` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)

**Pull request merged**